File: project.cpp
#include <iostream>
#include <fstream>
#include <String>

#include <chrono>
using namespace std::chrono;
using std::ifstream;
using std::cout;
using std::string;

int size = 0;


bool testWord(string word, string* words) {
    int i = 0;
    while (i < size - 2) {
        if (words[i] == word) {
            return true;
        }
        i++;
    }
    return false;
}


int main() {
    std::cout << "Hello World!\n";

    ifstream infile("./dictionary.json");

    auto start = high_resolution_clock::now();

    string line = "";

    cout << "Loading Dictionary...\n";
    while (std::getline(infile, line)) {
        ++size;


    }

    infile.clear();
    infile.seekg(0);

    string* dict = new string[size];

    int index = 0;

    while (!infile.eof()) {

        string* newWord = new string;
        //cout << "Line " << index << "\n";
        std::getline(infile, *newWord);

        dict[index++] = *newWord;



    }




    

    string* allWords = new string[size-2];  // ignoring the first and last lines ('{' and '}')

    for (int i = 0; i < size; i++) {
        if (dict[i] == "}") {
            break;
        }
        if (dict[i] != "{") {
            string line = dict[i];
            int wordStart = line.find("\"");
            int wordEnd = line.find("\"", wordStart+1);
            string newWord = line.substr(wordStart + 1, wordEnd - wordStart -1);
            allWords[i - 1] = newWord;


        }
    }

    auto end = high_resolution_clock::now();


    cout << "Time to load dictionary: " << ((float)duration_cast<nanoseconds>(end - start).count()) / 1000000000 << " seconds\n\n\n";

    cout << "Testing for \"silkness\" (end)\n";
    start = high_resolution_clock::now();
    cout << testWord("silkness", allWords) << "\n";
    end = high_resolution_clock::now();
    cout << "Time to find: " << ((float)duration_cast<nanoseconds>(end - start).count()) / 1000000000 << " seconds\n\n";
    

    cout << "Testing for \"anopholes\" (beginning)\n";
    start = high_resolution_clock::now();
    cout << testWord("anopheles", allWords) << "\n";
    end = high_resolution_clock::now();
    cout << "Time to find: " << ((float)duration_cast<nanoseconds>(end - start).count()) / 1000000000 << " seconds\n";





    return 0;
}

File: project.scala
import scala.io.Source
import scala.io.StdIn



object dict {
    
    def main(args: Array[String]) : Unit = {

//////////////////////  Load Dictionary  //////////////////////////////////////////////////////////

        println("Hello, world\n\n\nLoading (unorganized) dictionary...")


        // Load our words and definitions from dictionary.json as a list of items
        val startLoad = System.nanoTime     // Begin timer to find time needed to load the data



        val dict = Source.fromFile("./dictionary.json").getLines().toList

        


        // Place each word (at the beginning of each line) into a list of words, this saves us a lot of time later, but makes the code a bit more awkward
        val allwords = readJSON(dict, List())




        // End the timer and output in terms of seconds
        val loadTime = (System.nanoTime - startLoad)
        val humanLoad: Float  = loadTime.toFloat / 1000000000

        println(s"System time to load dictionary.json: $humanLoad seconds\n")


//////////////////////  Use Dictionary  ///////////////////////////////////////////////////////////


        // Print a few test variables and the time used to print them
        println("testing if \"anopheles\" is in our dictionary (beginning)")
        testWord("anopheles", allwords)

        println("testing if \"enslave\" is in our dictionary (middle)")
        testWord("enslave", allwords)

        println("testing if \"silkness\" is in our dictionary (end)")
        testWord("silkness", allwords)

        println("testing ig \"zzzahhhh\" is in our dictionary (not)")
        testWord("zzzahhhh", allwords)




        // User input loops (really odd implementation since it's functional)
        checkLoop(allwords)
        
        defineLoop(allwords, dict)




        // Test all words from a .txt, line delimited
        /*val query = Source.fromFile("./SmallData.txt").getLines().toList
        for (term <- query){
            println(term)
            testWord(term, allwords)
        }*/

    }





//////////////////////  Primary Methods  //////////////////////////////////////////////////////////

//////////////////////  Load Dictionary  //////////////////////////////////////////////////////////

    def readJSON(dictionary: List[String], wordList: List[String]) : List[String] = {
        val jsonString: String = dictionary.head

        if (jsonString == "}" ){        // done if at closing bracket
            return wordList
        }


        val tail = dictionary.tail
        //println(dictionary)

        if (jsonString == "{") {        // Ignore the first, opening bracket
            return readJSON(tail, wordList)
        }


        else{
            val wordStart = jsonString.indexOf('\"') + 1  
            val wordEnd = (jsonString.substring(wordStart).indexOf('\"') + wordStart)

            val newWord = jsonString.substring(wordStart, wordEnd)
               
            val wordListPlus = newWord :: wordList
                
            return readJSON(tail, wordListPlus)
        }
    }


///////////////////////////////////////////////////////////////////////////////////////////////////




//////////////////////  High-order Methods

    def testWord(word: String, wordList: List[String]) : Unit = {
        val startAnswer = System.nanoTime
        val bool = hasWord(word, wordList)
        val answerTime = System.nanoTime - startAnswer
        val humanAnswer: Float = answerTime.toFloat / 1000000000
        println(bool)

        println(s"That took $humanAnswer seconds\n\n")
    }

    def defineWord(word: String, validWords: List[String],  dictionary: List[String]): String = {
        if (hasWord(word, validWords)){
            return getDefinition(dictionary, findDictIndex(word, validWords, dictionary))
        }
        else {
            return s"Word \"$word\" is not defined"
        }
    }





///////////////////     Sub-Methods

    def hasWord(searchFor: String, fromList: List[String]) : Boolean = {



        if (fromList.head == searchFor){
            return true
        }
        if (fromList.tail == List()){
            return false
        }
        else {
            return hasWord(searchFor, fromList.tail)
        }


    }

    
    
    def getDefinition(dictionary: List[String], index: Int): String = {
        val line: String = dictionary(index + 1)
        val wordStart = line.indexOf(':') + 1
        return (line.substring(wordStart + 2, line.length()-1).replace("\\n", "\n"))

    }




    def findDictIndex(word:String, validWords: List[String], dictionary: List[String]): Int = {
        return dictionary.size - validWords.indexOf(word) - 3
    }




//////////////////////  User Input Loops  /////////////////////////////////////////////////////////

    def checkLoop(wordList: List[String]): Unit ={
        while (true) {
            println("What word would you like to check? Enter \':q\' to quit")
            val word: String = StdIn.readLine().toLowerCase()
            if (word == ":q"){
                return 
            }
            testWord(word, wordList)


        }



    }




    def defineLoop(wordList: List[String], dictionary: List[String]): Unit ={
        while (true) {
            println("What word would you like to define? Enter \':q\' to quit")
            val word: String = StdIn.readLine().toLowerCase()
            if (word == ":q"){
                return 
            }
            val startTime = System.nanoTime
            println(defineWord(word, wordList, dictionary))
            val humanTime = (System.nanoTime - startTime).toFloat / 1000000000
            println(s"\nThat took $humanTime seconds\n\n")
        }
    }



///////////////////////////////////////////////////////////////////////////////////////////////////
 

}






